### PR TITLE
Quasi-static and dynamic Timoshenko Beam #10313

### LINKED ISSUE #10313: Euler-Bernoulli and Timoshenko beam element
## Rationale
Beam elements are useful in modeling long slender objects that are more susceptible to bending behavior. 

## Description
Added small strain, large strain and large rotation C0 timoshenko beam element (Euler-Bernoulli beam is a special case of timoshenko beam element)

## Impact
This does not change anything in the already existing code.

### PR DESCRIPTION
**Pull request by sveerara**
Strain and stress calculators and inertial force calculators have been added for the C0 Timoshenko beam element. 

Quasi-static and dynamic beam bending tests have been added. 

Nodal inertial force/torque calculators have been added and corresponding tests have been. 

closes #10313

### CODE REVIEW

**Comment by sveerara** [comment]
@bwspenc, all the tests are working fine now. Can you please review this PR when you have some time? Thanks!

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+A beam element (\ref{fig:beam}) is used to model the response of a structural element that is long in one dimension compared its cross-section. There are two popular formulation of beam elements:
+
+ !media media/tensor_mechanics_beam_representation.png width=40% margin-left=30px float=right id=fig:beam caption= Beam element with 2 nodes and 3 translational and 3 rotational degrees of freedom at each node.
+
+* Euler-Bernoulli beam element: used to model bending deformation in long and slender beams. The two main assumptions in this beam theory are that: (i) the beam cross-section is rigid and does not deform under the application of transverse or lateral loads, and (ii) the cross-section of the beam remains planar and normal to the deformed axis of the beam.
+
+* Timoshenko beam element \citep{timoshenko_correction_1921, timoshenko_transverse_1922}: used to model both shear and bending deformation in short and thick beams. The beam cross-section does not deform in this beam theory as well and it remains planar. But the cross-section need not be normal to the deformed axis of the beam. Euler-Bernoulli beam element can be derived as a special case of the Timoshenko beam element.
```
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: Add 'The' in front of the last sentence.

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+Therefore, a C0 timoshenko beam element is implemented in MOOSE. This element has two nodes and each node has 6 degrees of freedom (dof) - 3 translational and 3 rotational displacements. All the 12 dofs are considered to be independent and the variation of both translational and rotational displacements along the length of the beam are modeled using first order Lagrange shape functions. The independent rotational dofs at the nodes makes it easier to model the shear deformation which results in non-perpendicular cross-sections with respect to the beam axis.
```
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: Capitalize 'Timoshenko'.  Also, I think it's better to capitalize 'dof', but that's up to you.

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+\epsilon_{12}(x,y,z) = \frac{\partial u_1(x,y,z)}{\partial y} + \frac{\partial u_2(x,y,z)}{\partial x} = -\theta_3 + \frac{\partial u_2(x)}{\partial x} - \frac{\partial \theta_1(x)}{\partial x} z
+$$
+
+$$
+\epsilon_{13}(x,y,z) = \frac{\partial u_1(x,y,z)}{\partial z} + \frac{\partial u_3(x,y,z)}{\partial x} = \theta_2 + \frac{\partial u_3(x)}{\partial x} + \frac{\partial \theta_1(x)}{\partial x} y
+$$
+
+It should be noted here that using a linear interpolation of the rotational variables in the calculation of shear strain leads to shear locking of the beam. Using $\theta_1(x) = ({\theta_1}^0 + {\theta_1}^1)/2$ in the calculation of $\epsilon_{12}$ and $\epsilon_{13}$ ensures that the beam doesn't lock under shear deformation \citep{prathap_reduced_1982}.
+
+The above translational strain increments are functions of x, y and z. However, since the beam cross-section does not deform, these strain can be integrated over the cross-section to obtain strain increments as a function of only x.
```
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: strain->strains

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+$$
+
+$$
+\epsilon_{13}(x,y,z) = \frac{\partial u_1(x,y,z)}{\partial z} + \frac{\partial u_3(x,y,z)}{\partial x} + \frac{\partial u_1(x,y,z)}{\partial x} \frac{\partial u_1(x,y,z)}{\partial z} + \frac{\partial u_2(x,y,z)}{\partial x} \frac{\partial u_2(x,y,z)}{\partial z}
+$$
+
+Note that for the `large_strain` calculation, $A_y$, $A_z$ and $J$ are assumed to be zero for simplicity. This is again valid for symmetric cross-sections such as square, rectangular and circular cross-sections.
+
+## Stress calculation
+The axial and shear strain increments, and the rotational strain increments are passed onto to [ComputeBeamForces](/ComputeBeamForces.md) to calculate the force ($\Delta F$) and moment ($\Delta M$) increments using the linear elastic constitutive relations defined in [ComputeElasticityBeam](/ComputeElasticityBeam.md) as follows:
```
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: remove 'onto'

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+$$
+
+As can be seen from the equation for $\epsilon_{1}(x)$, it depends only on $\frac{\partial u_1}{\partial x}$ if $A_y$ and $A_z$ are zero. But $\epsilon_{2}(x)$ depends on both $\frac{\partial u_1}{\partial x}$ and $\theta_3$. Therefore, $F_2(x) \delta \epsilon_{2}(x)$ would contribute to the residual of both the translational dof in y direction and rotational dof in the z direction. Similarly, $F_3(x) \delta \epsilon_{3}(x)$ would contribute to the residual of both the translational dof in z direction and rotational dof in the y direction. This accounts for the coupling between the shear and bending behavior of the beam. The stress-divergence calculation is performed in [StressDivergenceTensorsBeam](/StressDivergenceTensorsBeam.md).
+
+## Dynamic beam
+There are two main ways to include the inertial effects for the beam. The first method is to assign a uniform density to the beam and calculate a consistent mass/inertia matrix for the beam, and the second method assumes the beam to be massless and adds point mass/inertia to the nodes at the end of the beam.
```
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: I think I'd reword the last part of this:
'the second method assumes that the mass of the beam is concentrated at the ends of the beam, and represents the mass of the beam using point mass/inertia at the nodes at the end of the beam.'

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+If $A_y$ and $A_z$ are non-zero, then there is coupling between $u_1$, $\theta_2$ and $\theta_3$, $u_2$ and $\theta_1$, and $u_3$ and $\theta_1$.
+
+### Point mass/inertia
+[NodalInertialForce](/NodalInertialForce.md) and [NodalInertialTorque](/NodalInertialTorque.md) are used to apply mass/inertia at a node. The mass and inertia terms contribute only to the residual of the node at which it is applied. Mass (m) behaves isotropically in all three coordinate directions resulting in the following $j^{th}$ translational nodal residual:
+$$
+R_j = m \ddot{u}_j
+$$
+
+[NodalInertialTorque](/NodalInertialTorque.md) takes 6 inertia components as input ($I_{xx}$, $I_{yy}$, $I_{zz}$, $I_{xy}$, $I_{xz}$, $I_{yz}$). If coordinate system in which the moment of inertia components are provided is different from the global coordinate system, then `x_orientation` and `y_orientation` need to be provided as input so that a transformation matrix from the local coordinate system to the global coordinate system can be calculated using which the moment of inertia tensor in the global coordinate system can then be obtained.
```
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: Add 'the' after 'If'.

Also, the last part of the last sentence doesn't really make sense (the 'which' seems extraneous).

File: docs/content/documentation/systems/Kernels/tensor_mechanics/InertialForceBeam.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# InertialForceBeam
```
Comment (bwspenc) [inline docs/content/documentation/systems/Kernels/tensor_mechanics/InertialForceBeam.md]: These files all need some descriptions of what the classes do.  These can be really short, and should all reference the very thorough documentation you provided in C0TimoshenkoBeam.md.

**Review by bwspenc — CHANGES_REQUESTED** [review]
This is as far as I've gotten so far. I'll keep working on this -- just wanted to give you feedback so you could fix a few minor things with the documentation while I'm going through the rest of it.  This is looking really good!

File: modules/tensor_mechanics/include/materials/ComputeElasticityBeam.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef COMPUTEELASTICITYBEAM_H
+#define COMPUTEELASTICITYBEAM_H
+
+#include "Material.h"
+
+/**
+ * ComputeElasticityBeam defines the elasticity vector for the beam element
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/ComputeElasticityBeam.h]: Since 'elasticity vector' isn't a standard thing, maybe better to explain that a little more.  I'd say something like:

ComputeElasticityBeam computes the equivalent of the elasticity tensor for the beam element, which are vectors of material translational and flexural stiffness.

File: modules/tensor_mechanics/include/nodalkernels/NodalInertialForce.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef NODALINERTIALFORCE_H
+#define NODALINERTIALFORCE_H
+
+#include "NodalKernel.h"
+
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/nodalkernels/NodalInertialForce.h]: Rename this `NodalTranslationalInertia`?

File: modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+/****************************************************************/
+
+#ifndef NODALINERTIALTORQUE_H
+#define NODALINERTIALTORQUE_H
+
+#include "NodalKernel.h"
+#include "RankTwoTensor.h"
+
+// Forward Declarations
+class NodalInertialTorque;
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h]: Rename `NodalRotationalInertia`?

File: modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+template <>
+InputParameters validParams<NodalInertialTorque>();
+
+/**
+ * Calculates the inertial force and mass proportional damping for nodal inertia
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h]: This description is for the translational version.

**Comment by sveerara** [comment]
@bwspenc I have added the eigenstrain classes and the eigenstrain transfer class. Please review the PR when you have time. Some test in Rattlesnake is failing. My commits don't change any existing classes so its probably not related to this PR.

File: modules/tensor_mechanics/include/transfers/VectorPostprocessorToAuxVar.h [hunk 12]
```diff
@@ -0,0 +1,57 @@
+//* This file is part of the MOOSE framework
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/transfers/VectorPostprocessorToAuxVar.h]: I'm looking at the other transfers to see if the naming convention for this matches those, and it seems that something like 'VectorPostprocessorTransfer' might be more consistent. None of the transfers have anything about what kind of thing they are transferring to in their name.

File: modules/tensor_mechanics/include/userobjects/VectorPostprocessorToInterpolator.h [hunk 13]
```diff
@@ -0,0 +1,60 @@
+//* This file is part of the MOOSE framework
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/VectorPostprocessorToInterpolator.h]: I think I would rename this to something like 'CSVInterpolator' or 'InterpolateFromCSV'. The CSV files happen to be generated by a VectorPostprocessor in your case, but they could really come from anywhere.

File: docs/content/documentation/systems/Materials/tensor_mechanics/ComputeFiniteBeamStrain.md [hunk 14]
```diff
@@ -0,0 +1,12 @@
+#ComputeFiniteBeamStrain
+
+!syntax description /Materials/ComputeFiniteBeamStrain
+
+##Description
+This class computes the rotation matrix that relates the global coordinate system to the current beam local coordinate system for large rotation scenarios in beam elements. Please refer to [C0TimoshenkoBeam](/C0TimoshenkoBeam.md) for details.
```
Comment (bwspenc) [inline docs/content/documentation/systems/Materials/tensor_mechanics/ComputeFiniteBeamStrain.md]: This description (and the one in the header for this class) makes it sound like it only computes the rotation matrix. It also computes the strain increments (inherited from the base class). Change the description to make that clear.

File: modules/combined/test/tests/beam_eigenstrain_transfer/master.i [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+  [./fixr3]
+    type = DirichletBC
+    variable = rot_z
+    boundary = 1
+    value = 0.0
+  [../]
+[]
+
+[Kernels]
+  [./solid_disp_x]
```
Comment (bwspenc) [inline modules/combined/test/tests/beam_eigenstrain_transfer/master.i]: We should really add an Action like we have for the other StressDivergence kernels so you don't need to set up all of these (especially since we have 6 rather than 3 components).

It looks like we don't have that for the Truss ones either. Come to think of it, can we get the truss formulation by using a subset of the beam code?  It would be good to consolidate the truss and beam capabilities as much as possible (as long as it doesn't make the truss input more complex).

**Review by bwspenc — CHANGES_REQUESTED** [review]

**Comment by bwspenc** [comment]
@sveerara It sounds like the documentation system will be down until March 22. We can either wait until then to merge this, or strip out the Markdown files from this PR. Apparently there will be some changes to the Markdown file format (which is why it is frozen), but I imagine that it will be pretty easy to convert your stuff over.  Since it's a little ways off, I'd be inclined to go ahead and merge it before then without the Markdown.

File: modules/tensor_mechanics/src/materials/ComputeElasticityBeam.C [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+InputParameters
+validParams<ComputeElasticityBeam>()
+{
+  InputParameters params = validParams<Material>();
+  params.addParam<FunctionName>(
+      "elasticity_prefactor",
+      "Optional function to use as a scalar prefactor on the elasticity vector for the beam.");
+  params.addCoupledVar("youngs_modulus", "Variable containing Youngs modulus");
+  params.addCoupledVar("shear_modulus", "Variable contiaining shear modulus");
+  params.addCoupledVar("shear_coefficient", "Reduction factor for the shear modulus");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeElasticityBeam.C]: I'd prefer "scale factor" over "reduction factor" in this description.  Also, I think I'd change the descriptions of those coupled variables to say something like "Young's modulus of the material. Can be supplied as either a number or a variable name." ( so it's less confusing to the typical user who will probably just supply a number rather than a variable).

File: modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromAuxVar.C [hunk 17]
```diff
@@ -0,0 +1,68 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ComputeBeamEigenstrainFromAuxVar.h"
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromAuxVar.C]: I think I'd re-name this "ComputeBeamEigenstrainFromVariable" because it could work with either an AuxVariable or a Variable. Also, change the parameter descriptions to reflect that fact as well (remove 'aux').

**Comment by sveerara** [comment]
@bwspenc I will remove the documentation files when I address your comments. I was thinking of creating an action that creates  StressDivergenceBeam objects and also creates the NewmarkVelAux, NewmarkAccelAux and InertialForceBeam if `dynamic_consistent_mass` is set to true. And then maybe add a option for the nodal forces/inertia.

I will look at the truss files and see how it can be consolidated with the beam files. The beam can perform the functions of a truss as well but it would be storing more material properties than what a truss requires. If memory is not a problem, then I can make a truss class that inherits from the beam classes and set the unwanted strains/forces/moments to zero. I can also create separate classes for the truss using the plug-and-play model (strain calculator, elasticity and force calculator). In seperate classes are created, only axial force and strain need to be stored so it might be more memory efficient. Please let me know which method is better.

File: modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromVPP.C [hunk 18]
```diff
@@ -0,0 +1,112 @@
+//* This file is part of the MOOSE framework
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromVPP.C]: Since it's not coming straight from a VPP, I think a name like "ComputeBeamEigenstrainFromCSVInterpolator" would be more clear (assuming you rename VectorPostprocessorToInterpolator to CSVInterpolator).

**Comment by bwspenc** [comment]
@sveerara I wonder if we could keep the current truss material model (since it's simpler and doesn't need the memory for as many properties), but make it so you can use a common Kernel for both the translational part of the beam formulation and the truss.

**Comment by sveerara** [comment]
@bwspenc and @acasagran, I have updated the PR.

**Comment by sveerara** [comment]
I will remove the documentation files and squash all the commits into one once the reviews are completed and the PR is ready to be merged.

File: modules/tensor_mechanics/test/tests/beam/action/tests [hunk 19]
```diff
@@ -0,0 +1,13 @@
+[Tests]
+  [./2_block_action]
+    test = Exodiff
```
Comment (acasagran) [inline modules/tensor_mechanics/test/tests/beam/action/tests]: This should be type = 'Exodiff' both here and below.

File: modules/tensor_mechanics/test/tests/beam/action/2_block_common.i [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+  [../]
+  [./disp_x_2]
+    type = PointValue
+    point = '1.0 1.0 0.0'
+    variable = disp_x
+  [../]
+[]
+
+[Outputs]
+  file_base = '2_block_common_out'
```
Comment (acasagran) [inline modules/tensor_mechanics/test/tests/beam/action/2_block_common.i]: This should be '2_block_out' if you want to compare to the previous test which it looks like from the tests file.

**Comment by sveerara** [comment]
This PR has too many files so I am breaking it into smaller PRs.
